### ISSUE TITLE
Post with UrlRequest produce "a bytes-like object is required, not 'str'" error

### ISSUE THREAD

**Issue by algorys** [kivy/kivy#5730]
### Versions

* Python: 3.5.2
* OS: Ubuntu 16.04 LTS
* Kivy: v1.10.0
* Kivy installation method: package

### Description

I'm trying to make an http request (POST) with Kivy, but I got following error: `Request error   : a bytes-like object is required, not 'str'`.

With `requests` I've got no problem, but I can't put this library in requirements... due to a conflict with Kivy apparently, when I use buildozer...

### Code and Logs

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
TODO
"""

import kivy
import requests

from kivy.app import App
from kivy.uix.boxlayout import BoxLayout

from kivy.network.urlrequest import UrlRequest

kivy.require('1.10.0')
backend_url = 'http://demo.alignak.net:5000'
# backend_url = 'http://127.0.0.1:5000'


class MyApp(BoxLayout, App):
    """
        TODO
    """

    def build(self):
        """
        Override build
        TODO
        """

        username = 'admin'  # self.username.text
        password = 'admin'  # self.password.text
        json = {'username': username, 'password': password}
        url = '/'.join([backend_url, 'login'])
        req = UrlRequest(
            url,
            req_body=json,
            on_failure=self.got_fail,
            on_error=self.got_error, debug=True,
            on_success=self.got_success,
            on_redirect=self.got_redirect
        )
        print(req)

        test = requests.post(
            url,
            json=json
        )
        print("With requests: ", test.content)

        return self

    def got_success(self, req, *args):
        print("got success {0}".format(req.url))

    def got_error(self, req, *args):
        print("got error {0}".format(req.url))
        print("Requests finished: ", req.is_finished)
        print("Request result   :", req.result)
        print("Request error   :", req.error)

    def got_fail(self, req, *args):
        print("got fail {0}".format(req.url))

    def got_redirect(self, req, *args):
        print("got redirect {0}".format(req.url))


if __name__ == '__main__':
    MyApp().run()
```

I get following output:

```
[INFO   ] [Logger      ] Record log in /home/algorys/.kivy/logs/kivy_18-05-05_50.txt
[INFO   ] [Kivy        ] v1.10.0
[INFO   ] [Python      ] v3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609]
[INFO   ] [Factory     ] 194 symbols loaded
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil, img_gif (img_ffpyplayer ignored)
[INFO   ] [OSC         ] using <multiprocessing> for socket
[INFO   ] [Window      ] Provider: sdl2(['window_egl_rpi'] ignored)
[INFO   ] [GL          ] Using the "OpenGL" graphics system
[INFO   ] [GL          ] Backend used <gl>
[INFO   ] [GL          ] OpenGL version <b'4.6.0 NVIDIA 390.48'>
[INFO   ] [GL          ] OpenGL vendor <b'NVIDIA Corporation'>
[INFO   ] [GL          ] OpenGL renderer <b'GeForce GTX 760M/PCIe/SSE2'>
[INFO   ] [GL          ] OpenGL parsed version: 4, 6
[INFO   ] [GL          ] Shading version <b'4.60 NVIDIA'>
[INFO   ] [GL          ] Texture max size <16384>
[INFO   ] [GL          ] Texture max units <32>
<UrlRequest(Thread-1, started daemon 140152401307392)>
[INFO   ] [Window      ] auto add sdl2 input provider
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
With requests:  b'{\n  "token": "1519825242473-b655aca9-1aa6-4981-bfef-239a6ae8f291"\n}\n'
got error http://demo.alignak.net:5000/login
Requests finished:  True
Request result   : None
Request error   : a bytes-like object is required, not 'str'
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event6
[INFO   ] [MTD         ] Read event from </dev/input/event6>
[INFO   ] [Base        ] Start application main loop
[WARNING] [MTD         ] Unable to open device "/dev/input/event6". Please ensure you have the appropriate permissions.
[INFO   ] [Base        ] Leaving application in progress...

Process finished with exit code 0
```

**Comment by algorys** [comment]
Solution seems to use a str for `req_body=`.

Maybe dict or bytes will be more logical for this data ?

Otherwise, issue can be closed.

**Comment by KeyWeeUsr** [comment]
**TL;DR wrong type passed to a function in CPython stdlib**

@algorys `bytes` yes, however `dict` not at all. Your particular endpoint might support `application/json`, however there are other formats, such as `text/xml`(or whatever) or `application/x-www-form-urlencoded`. We might do the conversion from <X> to <Y> just for the sake of throwing whatever a user might have at the `UrlRequest`, but it makes no sense. You don't want throwing `xml.etree.ElementTree` objects to that `_send_request()` method either.

Also, unless there is a *reproducible* example, it's kind of hard-ish to play with random error without traceback or at least a final error line number. For the future, try `http://httbin.org` :)

Here's your real error (py3):

```python
from http.client import HTTPConnection as H
for body in (str({'a': 1}), str({'a': 1}).encode('utf-8'), {'a': 1}):
    con = H('www.httpbin.org', 80)

    print(type(body), repr(body))
    con.request('POST', '/post', body=body)  # <- here

    con.getresponse().read(455)  # because I'm lazy, it's from content-length
    print('-' * 10)
```
